datalad/support/annexrepo.py

```diff
@@ -121,6 +121,7 @@ class AnnexRepo(GitRepo, RepoInterface):
     GIT_ANNEX_MIN_VERSION = '6.20180913'
     git_annex_version = None
     supports_direct_mode = None
+    repo_version_info = None
```

Comment:
PR looks good to me.  

minor notes/ideas: 

1. I wondered if a better fit name could be chosen for this class variable though. `repo_` prefix somewhat visually occludes with our `.repo` and possibly suggests that it might be about an instance, not overall annex functionality.  May be
- `annex_versions` (and thus `annex_versions["supported"]` and `annex_versions["upgradable"]`)
- `repository_versions` (closer to verbatim wording of `git annex version` output)

2. So we use `check_` for other two class level variables.  I do understand that `get_` is closer in what it does here, but may be to stay consistent with the other two, it should be e.g. `check_annex_versions`?

Reply:
Thanks for having a look.

>  1. I wondered if a better fit name could be chosen for this class variable though. `repo_` prefix somewhat visually occludes with our `.repo` and possibly suggests that it might be about an instance, not overall annex functionality.  May be
>
>    * `annex_versions` (and thus `annex_versions["supported"]` and `annex_versions["upgradable"]`)

I'm worried it wouldn't be clear whether we're talking about the version of git-annex or repository versions, even if a reader doesn't gloss over the "s".

>    * `repository_versions` (closer to verbatim wording of `git annex version` output)

This clears up the above problem.  There is some risk that the reader thinks it is the version of the current repo (they don't register the "s" or somehow think annex repo's can be in some hybrid state :>), which was my motivation for tacking on "_info".  But I'm not sure "repository_version_info" would clear up any real confusion, so I went with your shorter "repository_versions" suggestion.

> 1. So we use `check_` for other two class level variables.  I do understand that `get_` is closer in what it does here, but may be to stay consistent with the other two, it should be e.g. `check_annex_versions`?

Fair enough.

<details>
<summary>range-diff</summary>

```
1:  f547890d8 ! 1:  9932d5377 ENH: annexrepo: Add get_repo_version_info method
    @@ Metadata
     Author: Kyle Meyer <kyle@kyleam.com>
     
      ## Commit message ##
    -    ENH: annexrepo: Add get_repo_version_info method
    +    ENH: annexrepo: Add check_repository_versions method
     
         Provide a method for getting the supported and upgradable repository
         versions.  In the current release of git-annex (7.20190819), the
    @@ datalad/support/annexrepo.py: class AnnexRepo(GitRepo, RepoInterface):
          GIT_ANNEX_MIN_VERSION = '6.20180913'
          git_annex_version = None
          supports_direct_mode = None
    -+    repo_version_info = None
    ++    repository_versions = None
      
          # Class wide setting to allow insecure URLs. Used during testing, since
          # git annex 6.20180626 those will by default be not allowed for security
    @@ datalad/support/annexrepo.py: def check_direct_mode_support(cls):
              return cls.supports_direct_mode
      
     +    @classmethod
    -+    def get_repo_version_info(cls):
    ++    def check_repository_versions(cls):
     +        """Get information on supported and upgradable repository versions.
     +
    -+        The result is cached at `cls.repo_version_info`.
    ++        The result is cached at `cls.repository_versions`.
     +
     +        Returns
     +        -------
    @@ datalad/support/annexrepo.py: def check_direct_mode_support(cls):
     +          supported -> list of supported versions (int)
     +          upgradable -> list of upgradable versions (int)
     +        """
    -+        if cls.repo_version_info is None:
    ++        if cls.repository_versions is None:
     +            from datalad.cmd import Runner
     +            key_remap = {
     +                "supported repository versions": "supported",
     +                "upgrade supported from repository versions": "upgradable"}
     +            out, _ = Runner().run(["git", "annex", "version"])
     +            kvs = (ln.split(":", 1) for ln in out.splitlines())
    -+            cls.repo_version_info = {
    ++            cls.repository_versions = {
     +                key_remap[k]: list(map(int, v.strip().split()))
     +                for k, v in kvs if k in key_remap}
    -+        return cls.repo_version_info
    ++        return cls.repository_versions
     +
          @staticmethod
          def get_size_from_key(key):
2:  29bb64f11 ! 2:  5db9ee007 TST: skip_v6_or_later: Consider whether v5 is supported by git-annex
    @@ datalad/tests/utils.py: def dm_func(*args, **kwargs):
     +    from datalad.support.annexrepo import AnnexRepo
     +
          version = cfg.obtain("datalad.repo.version")
    -+    info = AnnexRepo.get_repo_version_info()
    ++    info = AnnexRepo.check_repository_versions()
      
     -    @skip_if(version >= 6, msg="Skip test in v6+ test run", method=method)
     +    @skip_if(version >= 6 or 5 not in info["supported"],
3:  fef8786e5 ! 3:  2e666e11f TST: annexrepo: Rework test to use get_repo_version_info()
    @@ Metadata
     Author: Kyle Meyer <kyle@kyleam.com>
     
      ## Commit message ##
    -    TST: annexrepo: Rework test to use get_repo_version_info()
    +    TST: annexrepo: Rework test to use check_repository_versions()
     
         When testing that AnnexRepo(..., version=6) has the intended effect,
         test_repo_version() asserts that we expect either a v6 or v7 repo
    @@ datalad/support/tests/test_annexrepo.py: def test_repo_version(path1, path2, pat
     -    # the one below to eq_(version, 7).
     -    assert_in(version, [6, 7])
     +    # Since git-annex 7.20181031, v6 repos upgrade to v7.
    -+    supported_versions = AnnexRepo.get_repo_version_info()["supported"]
    ++    supported_versions = AnnexRepo.check_repository_versions()["supported"]
     +    v6_lands_on = next(i for i in supported_versions if i >= 6)
     +    eq_(version, v6_lands_on)
      
4:  b8270590f = 4:  721eb8b82 TST: annexrepo: Don't assume v5 is the default repository version
```

</details>